File: dairytrack-api/MilkProduction/app/models/raw_milk.py
from app import db
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime

class RawMilk(db.Model):
    __tablename__ = 'raw_milks'
    id = db.Column(db.Integer, primary_key=True)
    cow_id = db.Column(db.Integer, db.ForeignKey('cows.id'), nullable=False)
    production_time = db.Column(db.DateTime, nullable=False, default=db.func.current_timestamp())
    expiration_time = db.Column(db.DateTime, nullable=False)
    volume_liters = db.Column(db.Numeric(5, 2), nullable=False)
    previous_volume = db.Column(db.Numeric(5, 2), nullable=True)
    status = db.Column(db.String(20), nullable=False)
    session = db.Column(db.Integer, nullable=False)
    daily_total_id = db.Column(db.Integer, db.ForeignKey('daily_milk_totals.id'), nullable=True)
    available_stocks = db.Column(db.Numeric(5, 2), nullable=False, default=0.0)
    created_at = db.Column(db.DateTime, nullable=False, default=db.func.current_timestamp())
    updated_at = db.Column(db.DateTime, default=db.func.current_timestamp(), onupdate=db.func.current_timestamp())
    is_expired = db.Column(db.Boolean, default=False)

    # Relationship with Cow
    cow = db.relationship('Cow', back_populates='raw_milks')

    # Relationship with DailyMilkTotal
    daily_total = db.relationship('DailyMilkTotal', back_populates='raw_milks')

    def to_dict(self):
        # Calculate time left until expiration
        current_time = datetime.utcnow()
        if self.expiration_time and self.expiration_time > current_time:
            time_left = self.expiration_time - current_time
            time_left_str = str(time_left).split('.')[0]  # Format without microseconds
        else:
            time_left_str = "Expired"

        return {
            'id': self.id,
            'cow_id': self.cow_id,
            'cow': self.cow.to_dict(include_raw_milks=False) if include_cow and self.cow else None,
            'production_time': self.production_time,
            'expiration_time': self.expiration_time,
            'volume_liters': self.volume_liters,
            'previous_volume': self.previous_volume,
            'status': self.status,
            'session': self.session,
            'daily_total_id': self.daily_total_id,
            'timeLeft': time_left_str,
            'available_stocks': self.volume_liters,
            'is_expired': self.is_expired,
            'created_at': self.created_at,
        }

    def __repr__(self):
        return f"RawMilk('{self.cow_id}', '{self.production_time}', 'Session {self.session}')"





@db.event.listens_for(RawMilk, 'before_insert')
def set_session_for_today(mapper, connection, target):
    # Pastikan production_time adalah objek datetime
    if isinstance(target.production_time, str):
        # Coba beberapa format datetime yang mungkin
        for fmt in ("%Y-%m-%dT%H:%M:%S", "%Y-%m-%dT%H:%M"):
            try:
                # Ubah string menjadi objek datetime
                target.production_time = datetime.strptime(target.production_time, fmt)
                break
            except ValueError:
                continue
        else:
            # Jika tidak ada format yang cocok, lemparkan error
            raise ValueError(f"Invalid datetime format: {target.production_time}")

    # Ambil tanggal produksi (hanya tanggal, tanpa waktu)
    production_date = target.production_time.date()

    # Hitung jumlah sesi yang sudah ada untuk cow_id pada hari itu
    session = Session(connection)
    session_count = session.query(func.count(RawMilk.id)).filter(
        func.date(RawMilk.production_time) == production_date,
        RawMilk.cow_id == target.cow_id
    ).scalar()

    # Set session berdasarkan jumlah sesi yang sudah ada + 1
    target.session = session_count + 1

File: dairytrack-mobile/lib/model/produktivitas/rawMilk.dart
import 'package:dairy_track/model/peternakan/cow.dart';
import 'package:dairy_track/model/produktivitas/dairyMilkTotal.dart';

class RawMilk {
  int? id;
  int cowId;
  DateTime productionTime;
  DateTime expirationTime;
  double volumeLiters;
  String? name;
  String? lactationPhase;
  bool lactationStatus; // Ubah tipe menjadi bool
  double? previousVolume;
  String status;
  int session;
  int? dailyTotalId;
  double availableStocks;
  DateTime createdAt;
  DateTime updatedAt;
  bool isExpired;

  // Relationship with Cow
  Cow? cow;

  // Relationship with DailyMilkTotal
  DailyMilkTotal? dailyTotal;

  RawMilk({
    this.id,
    required this.cowId,
    required this.productionTime,
    required this.expirationTime,
    required this.volumeLiters,
    this.name,
    this.lactationPhase,
    required this.lactationStatus, // Properti diperbarui
    this.previousVolume,
    required this.status,
    required this.session,
    this.dailyTotalId,
    required this.availableStocks,
    required this.createdAt,
    required this.updatedAt,
    required this.isExpired,
    this.cow,
    this.dailyTotal,
  });

  factory RawMilk.fromJson(Map<String, dynamic> json) {
    try {
      return RawMilk(
        id: json['id'] != null ? int.tryParse(json['id'].toString()) : null,
        cowId: json['cow_id'] != null
            ? int.tryParse(json['cow_id'].toString()) ?? 0
            : 0,
        name: json['name'] ?? 'Unknown',
        lactationPhase: json['lactation_phase'],
        productionTime: json['production_time'] != null
            ? DateTime.parse(json['production_time'])
            : DateTime.now(),
        expirationTime: json['expiration_time'] != null
            ? DateTime.parse(json['expiration_time'])
            : DateTime.now(),
        volumeLiters: json['volume_liters'] != null
            ? (json['volume_liters'] is String
                ? double.tryParse(json['volume_liters']) ?? 0.0
                : (json['volume_liters'] as num).toDouble())
            : 0.0,
        previousVolume: json['previous_volume'] != null
            ? (json['previous_volume'] is String
                ? double.tryParse(json['previous_volume'])
                : (json['previous_volume'] as num).toDouble())
            : null,
        lactationStatus: json['lactation_status'] == true, // Konversi ke bool
        status: json['status'] ?? 'Unknown',
        session: json['session'] != null
            ? int.tryParse(json['session'].toString()) ?? 0
            : 0,
        dailyTotalId: json['daily_total_id'] != null
            ? int.tryParse(json['daily_total_id'].toString())
            : null,
        availableStocks: json['available_stocks'] != null
            ? (json['available_stocks'] is String
                ? double.tryParse(json['available_stocks']) ?? 0.0
                : (json['available_stocks'] as num).toDouble())
            : 0.0,
        createdAt: json['created_at'] != null
            ? DateTime.parse(json['created_at'])
            : DateTime.now(),
        updatedAt: json['updated_at'] != null
            ? DateTime.parse(json['updated_at'])
            : DateTime.now(),
        isExpired: json['is_expired'] == true || json['is_expired'] == 'true',
        cow: json['cow'] != null ? Cow.fromJson(json['cow']) : null,
        dailyTotal: json['daily_total'] != null
            ? DailyMilkTotal.fromJson(json['daily_total'])
            : null,
      );
    } catch (e) {
      throw Exception(
          'Error parsing RawMilk JSON: $e. Data: ${json.toString()}');
    }
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'cow_id': cowId,
      'name': name,
      'lactation_phase': lactationPhase,
      'production_time': productionTime.toIso8601String(),
      'expiration_time': expirationTime.toIso8601String(),
      'volume_liters': volumeLiters,
      if (previousVolume != null) 'previous_volume': previousVolume,
      'lactation_status': lactationStatus, // Tetap sebagai bool
      'status': status,
      'session': session,
      if (dailyTotalId != null) 'daily_total_id': dailyTotalId,
      'available_stocks': availableStocks,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'is_expired': isExpired,
      if (cow != null) 'cow': cow!.toJson(),
      if (dailyTotal != null) 'daily_total': dailyTotal!.toJson(),
    };
  }
}
